### PR TITLE
6788: You can deselect your current view mode

### PR DESCRIPTION
**Pull request by ElderOrb**
WIP on making proper fix

### CODE REVIEW

**Comment by anshuman64** [comment]
https://highfidelity.fogbugz.com/f/cases/6788/You-can-deselect-your-current-view-mode

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2346/HighFidelity-Beta-PR11140-5c08a17a6bf5c04319aecb9deebc62f14cca2f7a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2346/HighFidelity-Beta-PR11140-5c08a17a6bf5c04319aecb9deebc62f14cca2f7a.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2350/HighFidelity-Beta-PR11140-733fbcb077434990b46a5ec42d79d701d13e9ed2.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2350/HighFidelity-Beta-PR11140-733fbcb077434990b46a5ec42d79d701d13e9ed2.dmg

**Comment by druiz17** [comment]
test this please

File: libraries/ui/src/VrMenu.cpp [hunk 1]
```diff
@@ -43,6 +43,41 @@ class MenuUserData : public QObjectUserData {
...
+                return QObject::eventFilter(o, e);
+            }
+
+        private:
+            QObject* _from;
+            QObject* _to;
+            QObject* _qmlParent;
+        };
+
+        new ExclusionGroupSetter(action, qmlObject, qmlParent);
```
Comment (druiz17) [inline libraries/ui/src/VrMenu.cpp]: I do not understand what this line is doing. It seems that you create a new object and just leave it hanging. Can you clarify what going on here.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2352/HighFidelity-Beta-PR11140-733fbcb077434990b46a5ec42d79d701d13e9ed2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2352/HighFidelity-Beta-PR11140-733fbcb077434990b46a5ec42d79d701d13e9ed2.exe

File: libraries/ui/src/VrMenu.cpp (hunk 1 above)
Comment (ElderOrb) [inline libraries/ui/src/VrMenu.cpp]: this object will become child of 'action' which is also QObject, so should be deletet at the same time as QAction.

Comment (ElderOrb) [inline libraries/ui/src/VrMenu.cpp]: I can move its functionality into 'MenuUserData' although it will require making it QObject-based

Comment (druiz17) [inline libraries/ui/src/VrMenu.cpp]: I don't see the object explicitly setting 'action' as the parent, so can you explicitly set the objects parent.

Comment (ElderOrb) [inline libraries/ui/src/VrMenu.cpp]: it gets set in constructor: 

ExclusionGroupSetter(QObject* from, QObject* to, QObject* qmlParent) : QObject(from)

from - is 'action'

Comment (druiz17) [inline libraries/ui/src/VrMenu.cpp]: Ok, I didn't see that.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2403/HighFidelity-Beta-PR11140-bd214172af4d8e899a559789ed72b8909775b29c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2403/HighFidelity-Beta-PR11140-bd214172af4d8e899a559789ed72b8909775b29c.exe

**Comment by ElderOrb** [comment]
updated version

**Comment by ElderOrb** [comment]
https://highfidelity.fogbugz.com/f/cases/edit/6788/You-can-deselect-your-current-view-mode

Steps to verify (applicable to both VR and desktop versions of tablet)

1. Open tablet
2. Go to Menu > View

Before fix: 
3. 'View mode' checkboxes visible as checkboxes
4. 'View mode' checkboxes can be unchecked

After fix: 
3. 'View mode' checkboxes are visible as radiobuttons
4. 'View mode' radiobuttons can't be unchecked

**Comment by anshuman64** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2772/HighFidelity-Beta-PR11140-81264905d71177bff9a238fc878f4ee1281c9a3d.dmg

**Comment by anshuman64** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2774/HighFidelity-Beta-PR11140-81264905d71177bff9a238fc878f4ee1281c9a3d.dmg

**Comment by anshuman64** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2823/HighFidelity-Beta-PR11140-81264905d71177bff9a238fc878f4ee1281c9a3d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2823/HighFidelity-Beta-PR11140-81264905d71177bff9a238fc878f4ee1281c9a3d.exe

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

'View mode' checkboxes are visible as radiobuttons **Pass**
'View mode' radiobuttons can't be unchecked **Pass**

File: interface/resources/qml/desktop/Desktop.qml [hunk 2]
```diff
@@ -50,7 +50,24 @@ FocusScope {
...
-    property var rootMenu: Menu { objectName: "rootMenu" }
+    property var rootMenu: Menu { 
+        objectName: "rootMenu" 
+
+        // for some reasons it is not possible to use just '({})' here as it gets empty when passed to TableRoot/DesktopRoot
+        property var exclusionGroupsByMenuItem : ListModel {}
+
+        function addExclusionGroup(menuItem, exclusionGroup)
+        {
+            console.debug('entering Desktop::addExclusionGroup: menuItem = ', menuItem, 'exclusionGroup = ', exclusionGroup);
```
Comment (druiz17) [inline interface/resources/qml/desktop/Desktop.qml]: Please remove debugging statements

File: interface/resources/qml/desktop/Desktop.qml [hunk 3]
```diff
@@ -50,7 +50,24 @@ FocusScope {
...
+            exclusionGroupsByMenuItem.append(
+                {
+                    'menuItem' : menuItem.toString(), 
+                    'exclusionGroup' : exclusionGroup.toString()
+                }
+            );
+            console.debug('Desktop::addExclusionGroup: ', exclusionGroup, 'rootMenu: ', this);
```
Comment (druiz17) [inline interface/resources/qml/desktop/Desktop.qml]: Same here, remove debugging statements

File: interface/resources/qml/hifi/tablet/TabletMenuStack.qml [hunk 4]
```diff
@@ -133,10 +157,24 @@ Item {
             }
         }
 
+        property Component exclusiveGroupMaker: Component {
+            ExclusiveGroup {
+                Component.onDestruction: {
+                    console.debug('ExclusionGroup destroyed: ', this)
```
Comment (druiz17) [inline interface/resources/qml/hifi/tablet/TabletMenuStack.qml]: remove debug statement

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2875/HighFidelity-Beta-PR11140-a0a43ebf263e2fcb5c74c6a83258fc0105a4516f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2875/HighFidelity-Beta-PR11140-a0a43ebf263e2fcb5c74c6a83258fc0105a4516f.exe

**Comment by ElderOrb** [comment]
updated

**Comment by murillodigital** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2877/HighFidelity-Beta-PR11140-a0a43ebf263e2fcb5c74c6a83258fc0105a4516f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2877/HighFidelity-Beta-PR11140-a0a43ebf263e2fcb5c74c6a83258fc0105a4516f.exe

**Pull request merged**